WMF Framework/Remote Notifications/Model/RemoteNotificationsModelController.swift

```diff
@@ -222,32 +200,39 @@ final class RemoteNotificationsModelController: NSObject {
 
     // MARK: Mark as read
 
-    public func markAsRead(_ notification: RemoteNotification) {
-        let moc = legacyBackgroundContext
-        moc.perform {
-            notification.isRead = true
-            self.save(moc: moc)
+    public func markAsReadOrUnread(notifications: Set<RemoteNotification>, shouldMarkRead: Bool, completion: @escaping () -> Void) {
+        
+        guard let viewContext = notifications.first?.managedObjectContext else {
+            completion()
+            return
         }
-    }
-
-    public func markAsRead(notificationWithID notificationID: String) {
-        processNotificationWithID(notificationID) { (notification) in
-            notification.isRead = true
+        
+        //it is assumed notifications are from view context. We want to only write on background context.
+        var keys: [String] = []
+        viewContext.performAndWait {
+            keys = notifications.compactMap { $0.key }
         }
```

Comment:
Since we already have the set of `RemoteNotification`s here and we're only interested in the stable `key` property on a `RemoteNotification`, is it necessary to access their managed object context here?

Would it be sufficient to replace this with just:
```swift
let keys = notifications.compactMap { $0.key }
```

Reply:
@staykids I might be overly cautious here but I was worried about accessing the notifications' `key` property from a non-main thread. Even though the notifications passed in originated from the view context, this method here is called from the `RemoteNotificationsMarkReadOrUnreadOperation`, which is likely not running on the main thread. Let me know if I'm thinking of this wrong though. I think another pattern I've seen referenced is to pass around `NSManagedObjectID`s instead, and use those to call `object(with objectID: NSManagedObjectID)` on a background context to pull background notifications and flip their `isRead` flag. That might be a more clear and common way to do this.

Reply:
@tonisevener If the objective is Core Data thread safety in the operation, maybe the best solution is just avoiding dealing with passing around the `NSManagedObject` `RemoteNotification`s as much as possible in it. What if instead of passing a `Set<RemoteNotification>` to the controller at the beginning of the process in the view model, we just pass a `Set<String>` of `RemoteNotification.key`s? This functionally provides what the operation needs (a bunch of notification keys to operate with) without having to do any `NSManagedObjectIDs` or cross-context object access dancing.

Then the sequence looks like: main thread provides a bunch of `String`s to the remote notifications controller (from Notifications Center view model, something like `remoteNotificationsController.markAsReadOrUnread(notificationKeys: Set(notificationKeys), shouldMarkRead: shouldMarkRead)`), controller spins up an operation that operates on non-main thread, operation creates a new managed object context to operate with on that non-main thread, and just uses those non-`NSManagedObject`, non-context bound `key` Strings to fetch with.

I'm certainly not an expert in this area though. If you'd prefer to leave it as-is lemme know and I can merge.

Reply:
@staykids that is a good idea and much simpler, I like it. I'm working towards that the api controller also expects to be fed notifications in the operation, where it extracts the notification `id` and `wiki` properties. So there was already a potential threading bug there that this surfaced. Just giving you a heads up that this will be slightly more complicated than just switching over to `Set<String>` in those spots. I'll keep working at it but might not get this in by EOD.

Reply:
> where it extracts the notification `id` and `wiki` properties.  ... Just giving you a heads up that this will be slightly more complicated than just switching over to `Set<String>` in those spots.

@tonisevener If it's helpful at all: if you're looking for something to simplify all call sites and pass around the pertinent data, maybe a new type could help. This would probably provide better context to the method parameters and also keep the call sites simple.

```swift
extension RemoteNotification {
    public typealias Identifiers = (key: String?, id: String?, wiki: String?)

    public var identifiers: Identifiers {
        return (key, id, wiki)
    }
}
```
	
Unfortunately, I think you run into hashable limitations for these tuples to be Sets, so maybe a plain ole struct would do.

```swift
extension RemoteNotification {
	public struct Identifiers: Hashable {
		let key: String?
		let id: String?
		let wiki: String?
	}

	public var identifiers: Identifiers {
		return Identifiers(key: key, id: id, wiki: wiki)
	}
}
```

Maybe there's a more apt name for the struct, best I could think of quickly.

> I'll keep working at it but might not get this in by EOD.

No worries at all on the timeline.

Reply:
@staykids thanks for the tip! I basically used what you suggested, but used the name `IdentifierGroup`. I don't love it, but it helps that it's a singular form since I'm passing arrays and sets of this type around, and `Identifiers` having a property of `id` felt weird as well.